compiler/luci/pass/src/QuantizeOnnxDequantizeLinearPass.cpp

```diff
@@ -0,0 +1,543 @@
+
+public:
+  void apply(void)
+  {
+    // The final const's dtype is the same with input_dtype by default
+    auto const_dtype = _p.input->dtype();
+    if (const_dtype == loco::DataType::U8)
+    {
+      // Onnx does not support int4 as of writing. We assume uint8
+      // tensor is actually int4 or uint4 if values are within [0,15]
+      if (value_range<loco::DataType::U8>(_p.input, {0, 15}))
+      {
+        if (value_range<loco::DataType::U8>(_p.zerop, {8, 8}))
+        {
+          const_dtype = loco::DataType::S4;
+        }
+        else if (value_range<loco::DataType::U8>(_p.zerop, {0, 15}))
+        {
+          const_dtype = loco::DataType::U4;
+        }
+      }
+    }
+
+    luci::CircleConst *quant_const = nullptr;
+    switch (const_dtype)
+    {
+      case loco::DataType::S4:
+        quant_const = gen_s4_quant();
+        break;
+      case loco::DataType::U4:
+        quant_const = gen_u4_quant();
+        break;
+      case loco::DataType::U8:
+        quant_const = gen_u8_quant();
+        break;
+      default:
+        throw std::runtime_error("Unsupported quantized dtype");
+    }
+
+    assert(quant_const); // FIX_ME_UNLESS
+
+    // set origin
+    std::vector<std::shared_ptr<luci::CircleNodeOrigin>> origin_vec{
+      luci::get_origin(_p.dequantize), luci::get_origin(_p.input), luci::get_origin(_p.scale),
+      luci::get_origin(_p.zerop)};
+
+    luci::add_origin(quant_const, luci::composite_origin(origin_vec));
+
+    replace(_p.custom_out).with(quant_const);
+  }
+
+private:
+  luci::CircleConst *gen_s4_quant(void)
+  {
+    assert(_p.input->dtype() == loco::DataType::U8);      // FIX_CALLER_UNLESS
+    assert(_p.scale->dtype() == loco::DataType::FLOAT32); // FIX_CALLER_UNLESS
+    assert(_p.zerop->dtype() == loco::DataType::U8);      // FIX_CALLER_UNLESS
+
+    auto quantized_node = _p.dequantize->graph()->nodes()->create<luci::CircleConst>();
+    quantized_node->dtype(loco::DataType::S4);
+    quantized_node->rank(_p.input->rank());
+    for (uint32_t i = 0; i < _p.input->rank(); ++i)
+    {
+      quantized_node->dim(i) = _p.input->dim(i);
+    }
+    quantized_node->shape_status(luci::ShapeStatus::VALID);
+
+    // Create S4 CircleConst
+    // NOTE S4 is saved as S8 in luci::CircleConst
+    const auto num_elems = _p.input->size<loco::DataType::U8>();
+    quantized_node->size<loco::DataType::S4>(num_elems);
+    for (uint32_t i = 0; i < num_elems; i++)
+    {
+      uint8_t u8_val = _p.input->at<loco::DataType::U8>(i);
+      assert(u8_val <= 15); // FIX_CALLER_UNLESS
+      quantized_node->at<loco::DataType::S4>(i) = static_cast<int8_t>(u8_val) - 8;
+    }
+
+    auto qparam = std::make_unique<luci::CircleQuantParam>();
+    {
+      const std::vector<float> scale_vector = get_scales(_p.scale);
+      const std::vector<int64_t> zerop_vector = get_zerops<loco::DataType::U8>(_p.zerop);
+
+      for (auto z : zerop_vector)
+        assert(z == 8); // FIX_CALLER_UNLESS
```

Comment:
Release mode build fails because `z` is not used. I'll remove this line in the following PR.